### PR TITLE
Store attributes in the `ParseProgram` declaration types.

### PR DESCRIPTION
**Pull request by JoshuaBatty**
This PR carries over any attributes that were parsed and stored in `sway_ast::Module` into the `ParseProgram` decl types. 

My primary motivation for adding this was so the language server could access:
1. the attribute tokens so we can assign correct semantic information to them for correct syntax highlighting
2. allow access to the doc attribute so we can provide in-editor documentation pop-ups for hover requests. 

See `sway-lsp/src/utils/attributes.rs` mod for the intended API for accessing these tokens. 

I did notice that in #1488 @ortho mentioned the next step would be to pass the attributes into the AST. Feel free to let me know if you had another idea for how these would be stored. I'm converting the `AttributeDecl` into the `Attribute` type below which basically throws away bracket & punctuation tokens and retains each of the `Ident`s. Then the `AttributesMap` is what is stored in the AST. Similar to how it worked before but now instead of having the map just store the vec of args, it also stores the attribute name. 

```rust
pub enum AttributeKind {
    Doc,
    Storage,
}

pub struct Attribute {
    pub name: Ident,
    pub args: Vec<Ident>,
}

pub type AttributesMap = HashMap<AttributeKind, Vec<Attribute>>;
```

### CODE REVIEW

File: sway-core/src/convert_parse_tree.rs [hunk 1]
```diff
@@ -640,20 +684,29 @@ fn get_attributed_purity(
...
-                    _otherwise => {
-                        return Err(ec.error(ConvertParseTreeError::InvalidAttributeArgument {
-                            attribute: "storage".to_owned(),
-                            span: arg.span(),
-                        }));
+    match attributes.get(&AttributeKind::Storage) {
+        Some(attrs) if !attrs.is_empty() => {
+            for attr in attrs {
+                if !attr.args.is_empty() {
+                    for arg in &attr.args {
```
[the remark this answers is not in the document]
Comment (JoshuaBatty) [reply sway-core/src/convert_parse_tree.rs:691]: Nice, thanks! I'll try and adopt this practice going forward. Will change now.

File: sway-core/src/convert_parse_tree.rs [hunk 2]
```diff
@@ -640,9 +684,13 @@ fn get_attributed_purity(
...
         }
     };
-    match attributes.get(STORAGE_PURITY_ATTRIBUTE_NAME) {
-        Some(args) if !args.is_empty() => {
-            for arg in args {
+    match attributes.get(&AttributeKind::Storage) {
+        Some(attrs) if !attrs.is_empty() => {
+            for arg in attrs
+                .iter()
+                .filter(|attr| !attr.args.is_empty())
```
Comment (otrho) [inline sway-core/src/convert_parse_tree.rs:691]: This `filter` probably isn't necessary anyway, since `flat_map` would just skip it, I assume.

**Review by otrho — APPROVED** [review]

**Review by eureka-cpu — COMMENTED** [review]
lgtm, was there an issue that could be linked to this?

**Comment by JoshuaBatty** [comment]
> lgtm, was there an issue that could be linked to this PR?

It's needed to unblock #2260.

**Review by eureka-cpu — APPROVED** [review]

**Review by otrho — APPROVED** [review]

**Pull request merged**